src/Core/Content/Product/SalesChannel/Listing/Processor/PagingListingProcessor.php

```diff
@@ -59,18 +63,28 @@ private function getLimit(Criteria $criteria, SalesChannelContext $context, Requ
         $limit = $request->query->has('limit') ? $request->query->getInt('limit') : null;
         $limit = $request->request->has('limit') ? $request->request->getInt('limit') : $limit;
 
-        // request > criteria > config
+        // Priority 1: Request parameter (body > query)
         if ($limit > 0) {
             return min($limit, $this->maxLimit);
         }
 
-        if ($criteria->getLimit() !== null && $criteria->getLimit() > 0) {
-            return min($criteria->getLimit(), $this->maxLimit);
+        // Priority 2: Criteria limit (unless max limit was already applied)
+        $limit = null;
+        if (!$criteria->hasState(RequestCriteriaBuilder::STATE_MAX_LIMIT_USED)) {
+            $limit = $criteria->getLimit();
```

Comment:
Take me a while to understand this scenario, so you ignore criteria.limit if maxLimit was being used, then i'd name it `STATE_FALLBACK_LIMIT_USED`

```
$fallbackLimitUsed = $criteria->hasState(RequestCriteriaBuilder::STATE_MAX_LIMIT_USED);

if ($fallbackLimitUsed || $limit === null || $limit <= 0) {
   $limit = $this->config->getInt('core.listing.productsPerPage', $context->getSalesChannelId());
}
```

But it's optional I think